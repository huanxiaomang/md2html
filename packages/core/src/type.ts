export interface VNode {
    type: Tags;
    children: (VNode | Text)[];
}

export interface Text {
    type: Tags.TEXT;
    text: string;
}

export enum Tags {
    H1 = 'h1',
    H2 = 'h2',
    H3 = 'h3',
    H4 = 'h4',
    H5 = 'h5',
    H6 = 'h6',
    P = 'p',
    LI = 'li',
    UL = 'ul',
    OL = 'ol',
    BLOCKQUOTE = 'blockquote',
    CODE = 'code',
    B = 'b',
    IMG = 'img',
    A = 'a',
    HR = 'hr',
    TEXT = 'text',
}