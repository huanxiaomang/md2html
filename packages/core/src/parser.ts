import { Tags, VNode } from "./type";

const reg_mark = /^(.+?)\s/;
const reg_sharp = /^\#/;
const reg_heading = /^(#{1,6}\s+)/;
const reg_blockQuote = /^(\>\s+)/;
const reg_unorderedList = /^((\*|-){1}\s+)/;
const reg_image = /\!\[(.*?)\]\((.*?)\)/g;
const reg_strongText = /\*{2}(.*?)\*{2}/g;
const reg_codeLine = /\`{1}(.*?)\`{1}/g;

export function parser(markdown: string): VNode[] {
    const reader = new Reader(markdown);
    const vnodes: VNode[] = reader.run();
    return vnodes;
}

class Parser {
    constructor() { }



    public parser(line: string): VNode {
        const advance = (step: number) => {
            line = line.slice(step);
        }

        if (reg_heading.test(line)) {

            const text = line.split('# ')[1];
            const first = line.indexOf('#');
            let end = first;
            while (line[end += 1] === '#') { }
            const level = end - first;
            return this.createHeading(text, level as any);

        }

        if (reg_strongText.test(line)) {
            reg_strongText.exec(line);

            const text = reg_strongText.exec(line)[1];
            return this.createStrongText(text);

        }
    }

    private createHeading(text: string, level: 1 | 2 | 3 | 4 | 5 | 6): VNode {

        return {
            type: Tags[`H${level}`],
            children: [{
                type: Tags.TEXT,
                text: text
            }]
        };
    }

    private createStrongText(text: string): VNode {
        return {
            type: Tags.B,
            children: [{
                type: Tags.TEXT,
                text: text
            }]
        };
    }

    public isEmptyLine(line: string): boolean {
        return line.length === 0;
    }


}

class Reader {
    private lines: string[];
    private VNodes: VNode[] = [];
    constructor(private text: string) {
        this.lines = this.splitLines(text);
    }

    private splitLines(text: string): string[] {
        return text.split('\n');
    }

    public run() {
        const { parser, isEmptyLine } = new Parser();

        while (this.lines.length > 0) {
            const currentLine = this.lines.shift();
            if (isEmptyLine(currentLine)) continue;

            const vNode: VNode = parser(currentLine);

            this.VNodes.push(vNode);
        }

        return this.VNodes;
    }
}