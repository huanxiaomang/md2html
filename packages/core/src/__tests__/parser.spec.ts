import { describe, expect, it } from "vitest";
import { parser } from "../parser";
import { Tags } from "../type";

describe("Parser", () => {
    describe('h1~h6', () => {
        it('# 标题1', () => {
            const result = parser('# 标题1');
            expect(result).toStrictEqual([{
                type: Tags.H1,
                children: [{ type: 'text', text: '标题1' }]
            }]);
        });

        it('一堆标题', () => {
            const result = parser(`# 标题1

## 标题2

### 标题3

#### 标题4

##### 标题5

###### 标题6`);

            expect(result).toMatchSnapshot();
        });
    });

    // describe('text', () => {
    //     it('text', () => {
    //         const result = parser('你好！');
    //         expect(result).toStrictEqual([{
    //             type: Tags.TEXT,
    //             text: '你好！'
    //         }]);
    //     });

    // });
    describe('strongText', () => {
        it('single strongText', () => {
            const result = parser('**粗体**');
            expect(result).toStrictEqual([{
                type: Tags.B,
                children: [{ type: 'text', text: '粗体' }]
            }]);
        });
        // it('multi strongText', () => {
        //     const result = parser('**粗体1** **粗体2** **粗体3**');
        //     expect(result).toStrictEqual([{
        //         type: Tags.B,
        //         children: [{ type: 'text', text: '粗体' }]
        //     }, {
        //         type: Tags.B,
        //         children: [{ type: 'text', text: '粗体1' }]
        //     }, {
        //         type: Tags.B,
        //         children: [{ type: 'text', text: '粗体3' }]
        //     }

        //     ]);
        // })
    });
    describe('strongText in h1', () => {
        it('text', () => {
            const result = parser('# 标题**1**');
            expect(result).toStrictEqual([{
                type: Tags.H1,
                children: [
                    { type: 'text', text: '标题' },
                    { type: Tags.B, children: [{ type: 'text', text: '1' }] }
                ]
            }]);
        });

    });
})