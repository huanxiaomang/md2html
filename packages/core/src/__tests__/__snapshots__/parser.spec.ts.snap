// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parser > h1~h6 > 一堆标题 1`] = `
[
  {
    "children": [
      {
        "text": "标题1",
        "type": "text",
      },
    ],
    "type": "h1",
  },
  {
    "children": [
      {
        "text": "标题2",
        "type": "text",
      },
    ],
    "type": "h2",
  },
  {
    "children": [
      {
        "text": "标题3",
        "type": "text",
      },
    ],
    "type": "h3",
  },
  {
    "children": [
      {
        "text": "标题4",
        "type": "text",
      },
    ],
    "type": "h4",
  },
  {
    "children": [
      {
        "text": "标题5",
        "type": "text",
      },
    ],
    "type": "h5",
  },
  {
    "children": [
      {
        "text": "标题6",
        "type": "text",
      },
    ],
    "type": "h6",
  },
]
`;
